experiments-core/src/foedus/ycsb/ycsb_load.cpp

```diff
@@ -94,71 +94,137 @@ ErrorStack ycsb_load_task(const proc::ProcArguments& args) {
+      uint32_t high = node * options.thread_.thread_count_per_group_ + ordinal, low = 0;
+      YcsbKey key;
+      while (remaining_user_inserts) {
+        key.build(high, low++);
+        user_keys.push_back(key);
+        user_inserted++;
+        --remaining_user_inserts;
       }
-      inserted++;
-      if (--remaining_inserts == 0) {
+      ASSERT_ND(remaining_user_inserts == 0);
+      if (!user_table_spread) {  // do it on behalf of only one worker
         break;
       }
     }
-    ASSERT_ND(remaining_inserts == 0);
-    if (!spread) {  // do it on behalf of only one worker
-      break;
+  }
+
+  if (extra_table_spread || node == 0) {
+    for (uint16_t ordinal = 0; ordinal < options.thread_.thread_count_per_group_; ordinal++) {
+      if (!extra_table_spread && ordinal != 0) {
+        break;
+      }
+      auto remaining_extra_inserts = extra_records_per_thread;
+      uint32_t high = node * options.thread_.thread_count_per_group_ + ordinal, low = 0;
+      YcsbKey key;
+      while (remaining_extra_inserts) {
+        key.build(high, low++);
+        extra_keys.push_back(key);
+        extra_inserted++;
+        --remaining_extra_inserts;
+      }
+      ASSERT_ND(remaining_extra_inserts == 0);
+      if (!extra_table_spread) {  // do it on behalf of only one worker
+        break;
+      }
     }
   }
 
   if (sort_load_keys) {
-    ASSERT_ND(keys.size());
-    std::sort(keys.begin(), keys.end());
-    for (auto &key : keys) {
-      YcsbRecord r('a');
-      COERCE_ERROR_CODE(xct_manager->begin_xct(context, xct::kSerializable));
-      COERCE_ERROR_CODE(user_table.insert_record(context, key.ptr(), key.size(), &r, sizeof(r)));
-      COERCE_ERROR_CODE(xct_manager->precommit_xct(context, &commit_epoch));
-    }
+    ASSERT_ND(user_keys.size());
+    std::sort(user_keys.begin(), user_keys.end());
+    std::sort(extra_keys.begin(), extra_keys.end());
   }
-  COERCE_ERROR_CODE(xct_manager->wait_for_commit(commit_epoch));
+
+  if (user_keys.size()) {
+    COERCE_ERROR(load_table(context, user_keys, &user_table));
+  }
+  if (extra_keys.size()) {
+    COERCE_ERROR(load_table(context, extra_keys, &extra_table));
+  }
+
   watch.stop();
-  LOG(INFO) << "[YCSB] Loaded " << inserted << " records in " << watch.elapsed_sec() << "s";
+  LOG(INFO) << "[YCSB] Loaded " << user_inserted << " user table records, "
+    << extra_inserted << " extra table records in " << watch.elapsed_sec() << "s";
+  return kRetOk;
+}
+
+ErrorStack YcsbLoadTask::load_table(
+  thread::Thread* context,
+  const std::vector<YcsbKey>& keys,
+#ifndef YCSB_HASH_STORAGE
+  storage::masstree::MasstreeStorage* table) {
+#else
+  storage::hash::HashStorage* table) {
+#endif
+  Engine* engine = context->get_engine();
+  auto* xct_manager = engine->get_xct_manager();
+  Epoch commit_epoch;
+
+  for (auto &key : keys) {
+    YcsbRecord r('a');
+    bool succeeded = false;
+    // With the try-retry rw-lock insertions during loading phase might
```

Comment:
Same as the previous comment for session timeout, probably this is not needed. Will check.